py/desispec/preproc.py

```diff
@@ -0,0 +1,275 @@
+        
+        #- Initial teststand data may be missing GAIN* keywords; don't crash
+        if 'GAIN'+amp in header:
+            gain = header['GAIN'+amp]          #- gain = electrons / ADU
+        else:
+            log.error('Missing keyword GAIN{}; using 1.0'.format(amp))
+            gain = 1.0
+
+        overscan, rdnoise = _overscan(rawimage[ii])
+        rdnoise *= gain
+        kk = _parse_sec_keyword(header['CCDSEC'+amp])
+        readnoise[kk] = rdnoise
+        
+        header['OVERSCN'+amp] = overscan
+        header['OBSRDN'+amp] = rdnoise
+        
+        #- Warn/error if measured readnoise is very different from expected
+        if 'RDNOISE'+amp in header:
+            expected_readnoise = header['RDNOISE'+amp]
+            if rdnoise < 0.5*expected_readnoise:
+                log.error('Amp {} measured readnoise {:.2f} < 0.5 * expected readnoise {:.2f}'.format(
+                    amp, rdnoise, expected_readnoise))
+            elif rdnoise < 0.9*expected_readnoise:
+                log.warn('Amp {} measured readnoise {:.2f} < 0.9 * expected readnoise {:.2f}'.format(
+                    amp, rdnoise, expected_readnoise))
+            elif rdnoise > 2.0*expected_readnoise:
+                log.error('Amp {} measured readnoise {:.2f} > 2 * expected readnoise {:.2f}'.format(
+                    amp, rdnoise, expected_readnoise))
+            elif rdnoise > 1.2*expected_readnoise:
+                log.warn('Amp {} measured readnoise {:.2f} > 1.2 * expected readnoise {:.2f}'.format(
+                    amp, rdnoise, expected_readnoise))
+        else:
+            log.warn('Expected readnoise keyword {} missing'.format('RDNOISE'+amp))
+        
+        #- subtract overscan from data region and apply gain
+        jj = _parse_sec_keyword(header['DATASEC'+amp])
+        data = rawimage[jj] - overscan
+        image[kk] = data*gain
+
+    #- Load mask
+    if mask is not False and mask is not None:
+        if mask is True:
+            mask = read_mask(camera=camera, dateobs=dateobs)
+        elif isinstance(mask, (str, unicode)):
+            mask = read_mask(filename=mask)
+    else:
+        mask = np.zeros(image.shape, dtype=np.int32)
+
+    if mask.shape != image.shape:
+        raise ValueError('shape mismatch mask {} != image {}'.format(mask.shape, image.shape))
+
+    #- Divide by pixflat image
+    if pixflat is not False and pixflat is not None:
+        if pixflat is True:
+            pixflat = read_pixflat(camera=camera, dateobs=dateobs)
+        elif isinstance(pixflat, (str, unicode)):
+            pixflat = read_pixflat(filename=pixflat)
+
+        if pixflat.shape != image.shape:
+            raise ValueError('shape mismatch pixflat {} != image {}'.format(pixflat.shape, image.shape))
+
+        if np.all(pixflat != 0.0):
+            image /= pixflat
+            readnoise /= pixflat
+        else:
+            good = (pixflat != 0.0)
+            image[good] /= pixflat[good]
```

Comment:
the pixflat generation code doesn't yet exist for DESI (it exists for BOSS, but with somewhat specific BOSS tuning).  That code should enforce that pixflat values are positive, but the belt-and-suspenders approach here could be replacing `good = (pixflat != 0.0)` with
```
Does that seem reasonable to you?
if np.any(pixflat < 0.0):
    log.warn('negative pixflat values')
good = (pixflat > 0.0)
```